### PR TITLE
fix(plugin-meetings): fix auto rejoin after meeting ended or in left …

### PR DESCRIPTION
**Pull request by arun3528**
Fixes https://jira-eng-gpk2.cisco.com/jira/browse/SPARK-199688
---

Make sure to have followed the [contributing guidelines](https://github.com/webex/webex-js-sdk/blob/master/CONTRIBUTING.md#submitting-a-pull-request) before submitting.

### CODE REVIEW

File: packages/node_modules/@webex/plugin-meetings/src/reconnection-manager/index.js [hunk 1]
```diff
@@ -280,7 +297,11 @@ export default class ReconnectionManager {
         Metrics.postEvent(reconnectMetric);
         if (reconnectError instanceof NeedsRejoinError) {
           // send call aborded event with catogery as expected as we are trying to rejoin
-          return this.rejoinMeeting(reconnectError.wasSharing);
+
+          if (this.autoRejoinEnabled) {
+            return this.rejoinMeeting(reconnectError.wasSharing);
+          }
+          throw new Error('Cannot re connect the meeting , please rejoin.');
```
Comment (mwegman) [inline packages/node_modules/@webex/plugin-meetings/src/reconnection-manager/index.js:304]: Do you want to keep the original `reconnectError` object as well?

Comment (arun3528) [reply packages/node_modules/@webex/plugin-meetings/src/reconnection-manager/index.js:304]: @mwegman  the parent function already throws that error to the developer

Comment (mwegman) [reply packages/node_modules/@webex/plugin-meetings/src/reconnection-manager/index.js:304]:
```suggestion
          throw new AggregateError([reconnectError], 'Cannot re connect the meeting , please rejoin.');
```
You could use the AggregateError

Comment (arun3528) [reply packages/node_modules/@webex/plugin-meetings/src/reconnection-manager/index.js]: @mwegman i did remove that as the needTOrejoin error should be sufficient

File: packages/node_modules/@webex/plugin-meetings/src/meeting/index.js [hunk 2]
```diff
@@ -318,6 +318,16 @@ const logRequest = (request, {header = '', success = '', failure = ''}) => {
  * @event meeting:self:lobbyWaiting
  * @instance
  * @type {Object}
+ * @property {Object} reason Reason why user left the meeting
```
Comment (marcin-bazyl) [inline packages/node_modules/@webex/plugin-meetings/src/meeting/index.js:321]: this doesn't look right - the reason should be in `meeting:self:left` event and not in `meeting:self:lobbyWaiting`

File: packages/node_modules/@webex/plugin-meetings/src/meeting/index.js [hunk 3]
```diff
@@ -318,6 +318,16 @@ const logRequest = (request, {header = '', success = '', failure = ''}) => {
...
+ * @memberof Meeting
+ */
+
+/**
+ * Meeting Self Left State
+ * Emitted when user leaves the meeting either by own/server kicking off the user/user not active for 40 seconds or more
```
Comment (marcin-bazyl) [inline packages/node_modules/@webex/plugin-meetings/src/meeting/index.js:327]: the comment says that this event is emitted when "user leaves the meeting either by own...." - does this mean that it should be emitted when client calls `leave()`? Currently the code doesn't do that.

File: packages/node_modules/@webex/plugin-meetings/src/meetings/index.js [hunk 4]
```diff
@@ -827,25 +829,18 @@ export default class Meetings extends WebexPlugin {
...
-                  feedbackId: meeting.correlationId,
-                  locusId: meeting.locusId,
-                  meetingId: meeting.locusInfo?.info?.webExMeetingId
-                }).then(() => this.destroy(meeting, MEETING_REMOVED_REASON.NO_MEETINGS_TO_SYNC));
-              }
-              else {
-                this.destroy(meeting, MEETING_REMOVED_REASON.NO_MEETINGS_TO_SYNC);
-              }
+              // destroy function also uploads logs
+              console.log('RECONNECT   destroy ', this.meeting);
```
Comment (marcin-bazyl) [inline packages/node_modules/@webex/plugin-meetings/src/meetings/index.js:840]: console log

File: packages/node_modules/@webex/plugin-meetings/src/reconnection-manager/index.js [hunk 5]
```diff
@@ -324,10 +344,19 @@ export default class ReconnectionManager {
...
+
+      throw new Error('Unable to rejoin a meeting already ended or inactive .');
+    }
+
     LoggerProxy.logger.info(`ReconnectionManager:index#executeReconnection --> Current state of meeting is ${this.meeting.state}`);
 
     // If the meeting state was left, no longer reconnect media
-    if (this.meeting.state === _LEFT_) {
+    // or the meeting was deleted in process of re connecting
+    if (this.meeting.state === _LEFT_ || this.meeting.state === _INACTIVE_) {
```
Comment (marcin-bazyl) [inline packages/node_modules/@webex/plugin-meetings/src/reconnection-manager/index.js:359]: I'm a bit confused by this. I thought the whole point of the new autoRejoin field in config was to rejoin when we are in "inactive" state, but here we don't allow rejoining in "inactive" state.

**Comment by marcin-bazyl** [comment]
This PR has quite a lot of "todo" comments. Are you planning to address them all before merging this PR? In the previous projects I worked on we only allowed todo comments in code if they were accompanied by a Jira reference - this way we made sure that the todos are actually in our backlog and will be eventually done. Otherwise the code base can easily be filled with lots of todos that never get done.

File: packages/node_modules/@webex/plugin-meetings/src/meeting/index.js (hunk 3 above)
Comment (arun3528) [reply packages/node_modules/@webex/plugin-meetings/src/meeting/index.js:327]: yes you are correct , I still need to add a JIRA when a user gets kicked off by the server manually

File: packages/node_modules/@webex/plugin-meetings/src/meeting/index.js (hunk 2 above)
Comment (arun3528) [reply packages/node_modules/@webex/plugin-meetings/src/meeting/index.js:321]: yep typo

File: packages/node_modules/@webex/plugin-meetings/src/reconnection-manager/index.js (hunk 5 above)
Comment (arun3528) [reply packages/node_modules/@webex/plugin-meetings/src/reconnection-manager/index.js:359]: I think this line wont be executed at all . Removing it 

Before line 347 was added reconnect was in progress but the meeting would have already ended before the reconnect promise would finish

Comment (arun3528) [reply packages/node_modules/@webex/plugin-meetings/src/reconnection-manager/index.js:359]: The rejoin toggle makes sure that you need to trigger the rejoin compared to SDK doing a rejoin for you

**Comment by arun3528** [comment]
@marcin-bazyl  have updated the PR with review comments and also added the JIRA for the refresh on SDk or on the web client

**Review by mwegman — APPROVED** [review]

File: packages/node_modules/@webex/plugin-meetings/src/meeting/index.js (hunk 2 above)
Comment (marcin-bazyl) [reply packages/node_modules/@webex/plugin-meetings/src/meeting/index.js:321]: looks like this PR got merged with this API typo not being fixed

**Pull request merged**